### ISSUE TITLE
wasm32-unknown-unknown: wasm-ld fails to relocate drop_in_place entry in some trait object vtables

### ISSUE THREAD

**Issue by zylthinking** [rust-lang/rust#157209]
## Environment

- **Rust version:** `rustc 1.96.0 (ac68faa20 2026-05-25)`
- **Binary:** `wasm32-unknown-unknown`, `cdylib`, opt-level=s, no LTO, no PIC

## Summary

For some trait object vtables compiled to `wasm32-unknown-unknown`, `wasm-ld` fails to apply `R_WASM_TABLE_INDEX_SLEB` relocation to the **`drop_in_place`** entry. The entry retains the raw **function index** instead of being converted to a **table index**. This causes `call_indirect` during `Box<dyn Trait>::drop` to receive an index far beyond the table bounds.

**Runtime error:**
```
RuntimeError: table index is out of bounds; 
```
This error was reported following the AI's initial modification, which was based on the hypothesis that `JointorHandle` was improperly crossing the WASM boundary. To address this, the AI implemented an array-based storage strategy to keep the `JointorHandle` within the WASM memory space, passing only the array index across the WASM–JS–WASM interface.

I believe this effectively eliminates the suspicion that the issue was caused by crossing the boundary. This conclusion is supported by two key details:

1. **Refutation of the Boundary-Crossing Hypothesis:** The array-based implementation successfully localized the `JointorHandle` within WASM, yet the underlying instability persisted, effectively neutralizing the theory that boundary crossing was the root cause.
2. **Instruction Pointer Corruption:** In the original error report, the trace pointed to code segments that were not even being executed. Since the issue vanished after implementing `std::mem::forget(upstream)`, I posit that the problem was likely the result of the instruction pointer for the `drop` operation failing completely. This provides indirect evidence that the metadata or function pointers stored within the WASM memory were already corrupted or incorrect prior to execution.

## The reasons for considering the AI analysis correct are:

1. As demonstrated in the code, calling other virtual functions within dyn JointorHandle is Ok

2. The AI-led modification successfully resolved the reported error.

```rust
        impl Drop for JointorHandle {
            fn drop(&mut self) {
                if self.0.needs_manual_drop() {
                    unsafe {
                        let jointor: &dyn Jointor = &**self.0;
                        let layout = std::alloc::Layout::from_size_align(
                            std::mem::size_of_val(jointor),
                            std::mem::align_of_val(jointor),
                        )
                        .unwrap();
                        let ptr = jointor as *const dyn Jointor as *const u8 as *mut u8;
                        std::alloc::dealloc(ptr, layout);
                    }
                } else {
                    unsafe {
                        std::ptr::drop_in_place(&mut *self.0);
                    }
                }
            }
        }
```

**Comment by zylthinking** [comment]
This bug was analyzed by an AI, and since I lacked the necessary background knowledge to pinpoint the root cause, I relied on its analysis to resolve the issue. After the fix was implemented and the error was resolved—confirming that subsequent logs could print successfully—I requested the AI to revert the changes to reproduce the original crash. I then had it re-compile the Wasm module in that broken state and provide a detailed analysis of the failure. I have attached that analysis above.

**Comment by theemathas** [comment]
Do you have the original rust source code available publicly? It's not really feasible to figure out the issue without the source code. Just the resulting binary is not enough.

**Comment by zylthinking** [comment]
[wasm-vtable-repro.tar.gz](https://github.com/user-attachments/files/28532354/wasm-vtable-repro.tar.gz)
I have attached all the code in the attachment. A specific location is shown in the figure. Based on the AI's re-analysis, it appears there is more than one error; however, looking at the values, they don't seem like indices (I'm not sure if this is an issue with the AI). Regardless, I can confirm that by using forget, at least this specific location seems to be working normally now.
<img width="796" height="516" alt="Image" src="https://github.com/user-attachments/assets/120b1107-8d1e-4f95-8869-fadd62b859a4" />

**Comment by theemathas** [comment]
In `streamkit/src/context.rs`, in the `bridge_to_js` function, there's a `JointorHandle` argument. How was this argument constructed in javascript? The RuntimeError you're getting could plausibly be caused by a corrupted value of `JointorHandle`, not a miscompilation. (I think the LLM analysis is incorrect.)

**Comment by workingjubilee** [comment]
@zylthinking Please remove, from your initial post, all LLM-derived analysis, so that we know what the actual report is. You can move this analysis to any later comment if you think it is valuable.

**Comment by workingjubilee** [comment]
The initial input you gave to the AI in order to persuade it to generate this output contains information about the bug that you have not replicated into this post, like the circumstances you actually encountered it under. Was this under a test harness? Debug builds? Release builds? What was the actual `cargo` command you used to create the binary and then at what point was the error raised? During page loading? After?

This makes it an inadequate bug report, as the AI can make up its own metrics and claim the result of a Python script proves something. You need to have expertise enough to be able to catch it in a lie in order to use it for a task. But we do not need proof, merely a complete report of what your actual experience was.

The AI has also failed to notice that https://github.com/llvm/llvm-project/pull/104043 is not actually a merged PR, so it could not have possibly addressed any possible miscompilations. And that is depending on the assumption that it is indeed a miscompilation.

**Comment by zylthinking** [comment]
> In `streamkit/src/context.rs`, in the `bridge_to_js` function, there's a `JointorHandle` argument. How was this argument constructed in javascript? The RuntimeError you're getting could plausibly be caused by a corrupted value of `JointorHandle`, not a miscompilation. (I think the LLM analysis is incorrect.)

```vue
<template>
    <canvas ref="canvas"></canvas>
</template>

<script lang="ts">
import * as streamkit from "streamkit";
import { ref, onMounted } from "vue";

let canvas = ref(null);

await streamkit.default();
streamkit.formats_declare();
streamkit.modules_declare();
await streamkit.ws_prepare("ws://127.0.0.1:3000/ws");

let handles = streamkit.load_js(streamkit.ContainerType.Ws, "ws://127.0.0.1:3000/ws");
let handle = handles[0];
</script>

<script setup lang="ts">
onMounted(async () => {
    if (!canvas.value) {
        return;
    }
    const pipend = streamkit.bridge_to_js(handle, "wgpu://*", null, canvas.value);
    console.log(pipend);
});
</script>
```

> The initial input you gave to the AI in order to persuade it to generate this output contains information about the bug that you have not replicated into this post, like the circumstances you actually encountered it under. Was this under a test harness? Debug builds? Release builds? What was the actual `cargo` command you used to create the binary and then at what point was the error raised? During page loading? After?
> 
> This makes it an inadequate bug report, as the AI can make up its own metrics and claim the result of a Python script proves something. You need to have expertise enough to be able to catch it in a lie in order to use it for a task. But we do not need proof, merely a complete report of what your actual experience was.
> 
> The AI has also failed to notice that [llvm/llvm-project#104043](https://github.com/llvm/llvm-project/pull/104043) is not actually a merged PR, so it could not have possibly addressed any possible miscompilations. And that is depending on the assumption that it is indeed a miscompilation.

This is my command line:
```bash
sh: wasm-pack --version
wasm-pack 0.15.0

sh: cd ../streamkit/; wasm-pack build --target web; cd -; npx vite build --emptyOutDir &&  vite preview
```
After completing this, I opened the browser, loaded the target page (the corresponding JS code was attached in the post), and observed the output in the browser console.

**Comment by workingjubilee** [comment]
cool! thank you. I would have had no idea `vite` was involved at all.

probably will not actually be important for retracing the issue, but if it somehow is, then we don't have to guess.
